app/reports/invitation_uptake.rb

```diff
@@ -0,0 +1,35 @@
+class InvitationUptake
+  def run
+    CSV.generate do |csv|
+      csv << ['FCA Number', 'Registered Name', 'Invitation Sent', 'Invitation Accepted', 'Password Reset Sent']
+      query.all.each do |row|
+        csv << [
+          row.fca_number,
+          row.registered_name,
+          nice_date(row.invitation_sent_at),
+          nice_date(row.invitation_accepted_at),
+          nice_date(row.reset_password_sent_at)
+        ]
+      end
+    end
+  end
+
+  private
+
+  def query
+    User.where.not(invitation_sent_at: nil)
+      .joins(principal: [:firm])
+      .select(
+        'firms.fca_number',
+        'firms.registered_name',
+        'users.invitation_sent_at',
+        'users.invitation_accepted_at',
+        'users.reset_password_sent_at'
```

Comment:
I had a chat with @JiggyPete earlier about the `reset_password_sent_at` and I think it might be a red herring. I concur with your thoughts above on the extra fields. The existing tests are fairly comprehensive so adding more to follow the actual process should be relatively straightforward I think?

Remember as well though that Mark wants to know _who_ the firms are so just including a count is not enough. Prefixing as a TL;DR might be of benefit though if you can find a way to easily add it to the CSV (or I guess you could run a separate report that gives you that).